### DISCUSSION TITLE
How to define message types in client/service?

### DISCUSSION THREAD

**Discussion by ekkaiaaa** [ros2-rust/ros2_rust#304] [Q&A] [answered]
Hello,
I would like to know how to define messages for a ros2_rust client/service. I saw in the examples the use of a user-defined struct called `AddTwoInts`, but do I need to implement the `Message` and `Service` traits for every struct I want to use in the client and service nodes?

**Comment by ekkaiaaa** [comment]
Documentation says that `User code never needs to call this trait's method, much less implement this trait` about the `Service` trait 🤔

**Comment by esteve** [comment, marked as answer]
@ekkaiasmith sorry for the slow reply. Messages are defined the standard ROS way, that is, by creating `.msg` and `.srv` text files that are then converted into structures suitable for a client library. The `AddTwoInts` service messages are defined in the `example_interfaces` package: https://github.com/ros2/example_interfaces

For more info on how to create custom messages, you can follow the official documentation at https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Custom-ROS2-Interfaces.html
